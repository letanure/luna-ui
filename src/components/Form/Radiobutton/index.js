import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';

import FontAwesomeIcon from '@fortawesome/react-fontawesome';

/**
 * Radiobutton component.
 * E.g.:
 * ```html
 *  <Form.Radiobutton id="rd3" kind="success" label="funk" group="rd" value="3" />
 *  <Form.Radiobutton id="rd4" kind="danger" label="samba" group="rd" value="4" />
 * ```
 */
class RadiobuttonComponent extends React.PureComponent {
  render() {
    const { config, kind, value, id, label, group, className } = this.props;

    const Input = styled.input`
      visibility: hidden;
      &:checked + label {
        color: ${(props) => props.theme.form.radiobutton.types[kind]};
      }
      &::not(:checked) + label {
        color: ${(props) => props.theme.form.radiobutton.checkIcon};
      }
      &:disabled ~ label {
        cursor: not-allowed;
        color: ${(props) => props.theme.form.checkbox.disabledText};
      }
      &:disabled + label {
        color: ${(props) => props.theme.form.radiobutton.disabled};
        background-color: ${(props) => props.theme.form.radiobutton.disabled};
        border-color: ${(props) => props.theme.form.radiobutton.disabledBorder};
      }
    `;

    const Radiobutton = styled.label`
      width: 16px;
      height: 16px;
      padding: 2px;
      cursor: pointer;
      font-size: 10px;
      border-radius: 50%;
      left: 11px;
      position: absolute;
      transition: color .15s linear;
      color: ${(props) => props.theme.form.radiobutton.checkIcon};
      background-color: ${(props) => props.theme.form.radiobutton.checkIcon};
      border: 1px solid ${(props) => props.theme.form.radiobutton.types[kind]};
      & + label {
        position: absolute;
        left: 27px;
        cursor: pointer;
      }
    `;

    const RadiobuttonWraper = styled.div`
      height: 40px;
      padding: 11px;
      display: flex;
      position: relative;
    `;

    return (
      <RadiobuttonWraper className={className}>
        <Input readOnly id={id} name={group} type="radio" {...config} value={value} />
        <Radiobutton htmlFor={id}>
          <FontAwesomeIcon icon="circle" />
        </Radiobutton>
        <label htmlFor={id}>{label}</label>
      </RadiobuttonWraper>
    );
  }
}

RadiobuttonComponent.propTypes = {
  label: PropTypes.string,
  className: PropTypes.string,
  group: PropTypes.string.isRequired,
  id: PropTypes.string.isRequired,
  value: PropTypes.string.isRequired,
  kind: PropTypes.string.isRequired,
  config: PropTypes.object
};

RadiobuttonComponent.defaultProps = {
  label: '',
  className: '',
  kind: 'primary',
  config: {}
};

export default RadiobuttonComponent;
