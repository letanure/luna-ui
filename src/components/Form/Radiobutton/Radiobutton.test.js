import React from 'react';
import sinon from 'sinon';
import { mount } from 'enzyme';
import renderer from 'react-test-renderer';

import theme from '../../../config/defaultTheme';
import ThemeProvider from '../../Theme/index';

import Radiobutton from './index';

describe('<Radiobutton />', () => {
  let radiobuttonList;
  let onChange;

  beforeEach(() => {
    onChange = sinon.spy();

    radiobuttonList =
      <ThemeProvider theme={theme}>
        <Radiobutton id="rd3" kind="success" label="funk" group="rd" value="3" config={{ checked: true }} />
        <Radiobutton id="rd4" kind="danger" label="samba" group="rd" value="4" config={{ onChange }} />
      </ThemeProvider>;
  });

  test('To be checked', () => {
    const wrapper = mount(radiobuttonList);
    expect(wrapper.find('input').at(0)
      .getDOMNode().checked).toBeTruthy();
  });

  test('Call onChange', () => {
    const wrapper = mount(radiobuttonList);
    const radiobutton = () => wrapper.find('input').at(1);
    radiobutton().simulate('change');
    expect(onChange).toHaveProperty('callCount', 1);
  });

  test('toMatchSnapshot', () => {
    const tree = renderer
      .create(radiobuttonList)
      .toJSON();

    expect(tree).toMatchSnapshot();
  });
});
