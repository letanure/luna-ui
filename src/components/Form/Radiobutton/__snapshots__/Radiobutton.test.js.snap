// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Radiobutton /> toMatchSnapshot 1`] = `
<div>
  <div
    className="sc-jTzLTM paZTa"
  >
    <input
      checked={true}
      className="sc-gqjmRU iFwxyt"
      id="rd3"
      name="rd"
      readOnly={true}
      type="radio"
      value="3"
    />
    <label
      className="sc-VigVT hMeTgU"
      htmlFor="rd3"
    />
    <label
      htmlFor="rd3"
    >
      funk
    </label>
  </div>
  <div
    className="sc-cSHVUG jynBIs"
  >
    <input
      className="sc-fjdhpX iJThNM"
      id="rd4"
      name="rd"
      onChange={[Function]}
      readOnly={true}
      type="radio"
      value="4"
    />
    <label
      className="sc-jzJRlG cdiKJv"
      htmlFor="rd4"
    />
    <label
      htmlFor="rd4"
    >
      samba
    </label>
  </div>
</div>
`;
